pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,412 @@
+		}).Debug("finished syncing machineset")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	machineSet, err := c.machineSetsLister.MachineSets(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("machineset has been deleted")
+		// TODO: ensure deleted in remote cluster. If so, how are we going to figure out what cluster
+		// the machine set was from to load the correct kubeconfig secret?
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if machineSet.Spec.NodeType == cov1.NodeTypeMaster {
+		return c.syncMasterMachineSet(machineSet)
+	} //else {
+	// TODO:
+	//c.syncComputeMachineSet(machineSet)
+	//}
+	return nil
+}
+
+// Creates the cluster in the remote.
+func (c *Controller) syncMasterMachineSet(ms *cov1.MachineSet) error {
+	msLog := logging.WithMachineSet(c.logger, ms)
+	msLog.Debug("loaded machineset")
+
+	if !ms.Status.ClusterAPIInstalled {
+		msLog.Debugf(errorMsgClusterAPINotInstalled)
+		return fmt.Errorf(errorMsgClusterAPINotInstalled)
+	}
+
+	remoteClusterAPIClient, err := c.buildClients(ms)
+	if err != nil {
+		return err
+	}
+
+	// Lookup the CO cluster object for this machine set:
+	coCluster, err := c.client.ClusteroperatorV1alpha1().Clusters(ms.Namespace).Get(ms.Labels[controller.ClusterNameLabel], metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error looking up machineset's cluster: %v", err)
+	}
+	msLog = logging.WithCluster(msLog, coCluster)
+
+	// Create the Cluster object if it does not already exist:
+	rCluster, err := remoteClusterAPIClient.ClusterV1alpha1().Clusters(remoteClusterAPINamespace).Get(coCluster.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// Cluster does not exist, create it:
+		msLog.Infof("creating new cluster API object in remote cluster")
+		rCluster = buildClusterAPICluster(coCluster)
+		_, err := remoteClusterAPIClient.ClusterV1alpha1().Clusters(remoteClusterAPINamespace).Create(rCluster)
+		if err != nil {
+			return fmt.Errorf("error creating cluster object in remote cluster: %v", err)
+		}
+	} else if err != nil {
+		return err
+	} else {
+		// TODO: patch/update here?
+		msLog.Debugf("cluster API object already exists in remote cluster")
+	}
+
+	/*
+
+		// Check if this machine set exists in the remote cluster, create if not. We assume
+		// the same name, and the "kube-cluster" namespace.
+		remoteMachineSet, err := remoteClusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterAPINamespace).Get(ms.Name, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			// Machine set does not exist in remote cluster, create it:
+		} else if err == nil {
+			// Machine set exists in remote cluster, see if we need to update it:
+		}
+	*/
+
+	return nil
+}
+
+func buildClusterAPICluster(cluster *cov1.Cluster) *clusterapiv1.Cluster {
+	capiCluster := clusterapiv1.Cluster{}
+	capiCluster.Name = cluster.Name
+	capiCluster.Spec.ProviderConfig.Value = &runtime.RawExtension{
```

Comment:
I updated this based on https://github.com/openshift/cluster-operator/pull/179 (and swapped imports to `sig.k8s.io/cluster-api`). Hitting this error on the sync. :sos: @csrwng

```
time="2018-05-01T20:54:43Z" level=info msg="Error syncing master machine set: error creating cluster object in remote cluster: Cluster in version \"v1alpha1\" cannot be handled as a Cluster: v1alpha1.Cluster.Spec: v1alpha1.ClusterSpec.ProviderConfig: ReadString: expects \" or n, but found {, error found in #10 byte of ...|rConfig\":{\"value\":{\"|..., bigger context ...|ocks\":null},\"serviceDomain\":\"\"},\"providerConfig\":{\"value\":{\"kind\":\"Cluster\",\"apiVersion\":\"clusterope|..." controller=syncmachineset machineset="{default/abutcher-master-zp82t abutcher}"
```

Reply:
@abutcher completely forgot about that... that's because we haven't updated the cluster-api image on quay.io to the latest.

Reply:
@abutcher I've now updated the image to one built with the latest code. Please try again.